src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileRunCommand.cs

```diff
@@ -0,0 +1,193 @@
+        {
+            ParseArguments();
+
+            var services = ConfigureDefaultServices();
+            Host = services.GetRequiredService<IMvcRazorHost>();
+            var fileProvider = services.GetRequiredService<IRazorViewEngineFileProviderAccessor>().FileProvider;
+            CompilationService = services.GetRequiredService<ICompilationService>() as IRoslynCompilationService;
+            if (CompilationService == null)
+            {
+                Console.Error.WriteLine(
+                    $"An {typeof(ICompilationService)} of type {typeof(IRoslynCompilationService)} " +
+                    "is required for Razor precompilation.");
+            }
+
+            var result = CompileViews(fileProvider);
+            if (!result.Success)
+            {
+                foreach (var error in result.RazorErrors)
+                {
+                    Console.Error.WriteLine($"{error.Location.FilePath} ({error.Location.LineIndex}): {error.Message}");
+                }
+
+                foreach (var error in result.RoslynErrors)
+                {
+                    Console.Error.WriteLine(CSharpDiagnosticFormatter.Instance.Format(error));
+                }
+
+                return 1;
+            }
+
+            var outputPath = GetBuildOutputPath();
+            UpdateAssembly(outputPath, result);
+            return 0;
+        }
+
+        private bool ParseArguments()
+        {
+            if (ConfigureCompilationType.HasValue())
+            {
+                var type = Type.GetType(ConfigureCompilationType.Value());
+                if (type == null)
+                {
+                    Console.Error.WriteLine($"Unable to find type '{type}.");
+                    return false;
+                }
+
+                var configureMethod = type?.GetMethod(ConfigureMvcMethod, BindingFlags.Public | BindingFlags.Static);
+                if (configureMethod == null)
+                {
+                    Console.Error.WriteLine($"Could not find a method named {ConfigureMvcMethod} on {type}.");
+                    return false;
+                }
+
+                ConfigureMvcAction = (Action<IMvcBuilder>)configureMethod.CreateDelegate(
+                    typeof(Action<IMvcBuilder>),
+                    target: null);
+            }
+
+            return true;
+        }
+
+        private void UpdateAssembly(string outputPath, PrecompilationResult precompilationResult)
+        {
+            var host = new PeReader.DefaultHost();
+
+            var assemblyPath = Path.Combine(outputPath, Path.GetFileName(ProjectPath) + ".dll");
+            var assembly = host.LoadUnitFrom(assemblyPath) as IAssembly;
+            assembly = new MetadataDeepCopier(host).Copy(assembly);
+
+            var rewriters = new MetadataRewriter[]
+            {
+                new AddResourcesRewriter(host, precompilationResult.CompileOutputs),
+                new RemoveStrongNameRewriter(host)
```

Comment:
Other than the fact that we don't have the smarts to re-apply strong naming this is pretty clean.